src/frontend/src/scheduler/distributed/query.rs

```diff
@@ -233,19 +216,19 @@ impl QueryRunner {
                         tracing::trace!("Query {:?} has scheduled all of its stages that have table scan (iterator creation).", self.query.query_id);
                         self.hummock_snapshot_manager
                             .unpin_snapshot(self.epoch, self.query.query_id())
-                            .await?;
+                            .await;
                     }
 
                     if self.scheduled_stages_count == self.stage_executions.len() {
                         // Now all stages have been scheduled, send root stage info.
                         self.send_root_stage_info().await;
                     } else {
                         for parent in self.query.get_parents(&stage_id) {
-                            if self.all_children_scheduled(parent).await {
-                                self.stage_executions[parent].start().await.map_err(|e| {
-                                    error!("Failed to start stage: {}, reason: {:?}", stage_id, e);
-                                    e
-                                })?;
+                            if self.all_children_scheduled(parent).await
+                                // Do not schedule same stage twice.
+                                && self.stage_executions[parent].is_pending().await
```

Comment:
I don't think this solves the dual start problem. During the stage start pocess, it still may read pending state twice. If you don't want to handle dual start, one solution is to record whether we have started stage   in query runner, which is a little ugly to me.

Reply:
Here's my thought: Once we called `.start()`, the state must become `Started`. The query runner is single-task so I can not imagine why it will think the stage is `Pending` again after already strart it.